File: ios/PetitesBouchees/ViewModels/ProgressViewModel.swift
import SwiftData
import Foundation

struct CategoryProgress: Identifiable, Sendable {
    var id: String { category.rawValue }
    let category: FoodCategory
    let total: Int
    let tested: Int
}

struct AlertItem: Identifiable, Sendable {
    let id: PersistentIdentifier
    let foodName: String
    let hasDigestiveIssue: Bool
    let hasAllergicReaction: Bool
}

@Observable
@MainActor
final class ProgressViewModel {
    private let modelContext: ModelContext

    var totalFoods: Int = 0
    var testedFoods: Int = 0
    var categoryProgress: [CategoryProgress] = []
    var likedCount: Int = 0
    var neutralCount: Int = 0
    var dislikedCount: Int = 0
    var alerts: [AlertItem] = []

    var progressPercent: Double {
        guard totalFoods > 0 else { return 0 }
        return Double(testedFoods) / Double(totalFoods)
    }

    init(modelContext: ModelContext) {
        self.modelContext = modelContext
    }

    func fetchProgress() {
        let foodDescriptor = FetchDescriptor<Food>()
        let allFoods = (try? modelContext.fetch(foodDescriptor)) ?? []

        let entryDescriptor = FetchDescriptor<FoodEntry>()
        let allEntries = (try? modelContext.fetch(entryDescriptor)) ?? []

        let testedFoodIDs = Set(allEntries.compactMap { $0.food?.persistentModelID })

        totalFoods = allFoods.count
        testedFoods = testedFoodIDs.count

        categoryProgress = FoodCategory.allCases.map { cat in
            let catFoods = allFoods.filter { $0.category == cat }
            let catTested = catFoods.filter { testedFoodIDs.contains($0.persistentModelID) }.count
            return CategoryProgress(category: cat, total: catFoods.count, tested: catTested)
        }

        likedCount = allEntries.filter { $0.reaction == .liked }.count
        neutralCount = allEntries.filter { $0.reaction == .neutral }.count
        dislikedCount = allEntries.filter { $0.reaction == .disliked }.count

        alerts = allEntries
            .filter { $0.hasDigestiveIssue || $0.hasAllergicReaction }
            .compactMap { entry in
                guard let food = entry.food else { return nil }
                return AlertItem(
                    id: entry.persistentModelID,
                    foodName: food.name,
                    hasDigestiveIssue: entry.hasDigestiveIssue,
                    hasAllergicReaction: entry.hasAllergicReaction
                )
            }
    }
}

File: app/src/main/kotlin/com/petitesbouchees/app/ui/progress/ProgressViewModel.kt
package com.petitesbouchees.app.ui.progress

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.petitesbouchees.app.data.model.FoodCategory
import com.petitesbouchees.app.data.model.Reaction
import com.petitesbouchees.app.data.repository.FoodRepository
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn

data class CategoryProgress(
    val category: FoodCategory,
    val total: Int,
    val tested: Int
)

data class AlertItem(
    val foodName: String,
    val hasDigestiveIssue: Boolean,
    val hasAllergicReaction: Boolean
)

data class ProgressUiState(
    val totalFoods: Int = 0,
    val testedFoods: Int = 0,
    val categoryProgress: List<CategoryProgress> = emptyList(),
    val likedCount: Int = 0,
    val neutralCount: Int = 0,
    val dislikedCount: Int = 0,
    val alerts: List<AlertItem> = emptyList(),
    val isLoading: Boolean = true
) {
    val progressPercent: Float
        get() = if (totalFoods > 0) testedFoods.toFloat() / totalFoods else 0f
}

class ProgressViewModel(private val repository: FoodRepository) : ViewModel() {
    val uiState: StateFlow<ProgressUiState> = combine(
        repository.getAllFoods(),
        repository.getAllEntries(),
        repository.getEntriesWithIssues()
    ) { foods, entries, issueEntries ->
        val foodMap = foods.associateBy { it.id }
        val testedFoodIds = entries.map { it.foodId }.toSet()

        val categoryProgress = FoodCategory.entries.map { category ->
            val total = foods.count { it.category == category }
            val tested = entries.count { entry -> foodMap[entry.foodId]?.category == category }
            CategoryProgress(category, total, tested)
        }

        val alerts = issueEntries.mapNotNull { entry ->
            foodMap[entry.foodId]?.let { food ->
                AlertItem(food.name, entry.hasDigestiveIssue, entry.hasAllergicReaction)
            }
        }

        ProgressUiState(
            totalFoods = foods.size,
            testedFoods = testedFoodIds.size,
            categoryProgress = categoryProgress,
            likedCount = entries.count { it.reaction == Reaction.LIKED },
            neutralCount = entries.count { it.reaction == Reaction.NEUTRAL },
            dislikedCount = entries.count { it.reaction == Reaction.DISLIKED },
            alerts = alerts,
            isLoading = false
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), ProgressUiState())

    class Factory(private val repository: FoodRepository) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return ProgressViewModel(repository) as T
        }
    }
}
